includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -376,185 +376,233 @@ public function payment_fields() {
+		];
+		if ( $is_recurring ) {
+			$metadata['payment_type'] = 'recurring';
+		}
 
-				// Update saved payment method information with checkout values, as some saved methods might not have billing details.
-				if ( $is_saved_method ) {
-					try {
-						$this->customer_service->update_payment_method_with_billing_details_from_order( $payment_method, $order );
-					} catch ( Exception $e ) {
-						// If updating the payment method fails, log the error message but catch the error to avoid crashing the checkout flow.
-						Logger::log( 'Error when updating saved payment method: ' . $e->getMessage() );
-					}
-				}
+		// Determine the customer making the payment, create one if we don't have one already.
+		$customer_id = $this->customer_service->get_customer_id_by_user_id( $user->ID );
 
-				$metadata = [
-					'customer_name'  => $name,
-					'customer_email' => $email,
-					'site_url'       => esc_url( get_site_url() ),
-					'order_id'       => $order->get_id(),
-				];
-
-				// Create intention, try to confirm it & capture the charge (if 3DS is not required).
-				$intent = $this->payments_api_client->create_and_confirm_intention(
-					WC_Payments_Utils::prepare_amount( $amount, 'USD' ),
-					'usd',
-					$payment_method,
-					$customer_id,
-					$manual_capture,
-					$save_payment_method,
-					$metadata,
-					$this->get_level3_data_from_order( $order )
-				);
+		if ( null === $customer_id ) {
+			// Create a new customer.
+			$customer_id = $this->customer_service->create_customer_for_user( $user, $name, $email );
+		} else {
+			// Update the existing customer with the current details. In the event the old customer can't be
+			// found a new one is created, so we update the customer ID here as well.
+			$customer_id = $this->customer_service->update_customer_for_user( $customer_id, $user, $name, $email );
+		}
+
+		// Update saved payment method information with checkout values, as some saved methods might not have billing details.
+		if ( $is_saved_method ) {
+			try {
+				$this->customer_service->update_payment_method_with_billing_details_from_order( $payment_method, $order );
+			} catch ( Exception $e ) {
+				// If updating the payment method fails, log the error message but catch the error to avoid crashing the checkout flow.
+				Logger::log( 'Error when updating saved payment method: ' . $e->getMessage() );
+			}
+		}
 
-				$intent_id = $intent->get_id();
-				$status    = $intent->get_status();
+		$intent_failed = false;
 
-				if ( $save_payment_method && ( 'succeeded' === $status || 'requires_capture' === $status ) ) {
-					try {
-						$this->token_service->add_payment_method_to_user( $payment_method, wp_get_current_user() );
-					} catch ( Exception $e ) {
-						// If saving the token fails, log the error message but catch the error to avoid crashing the checkout flow.
-						Logger::log( 'Error when saving payment method: ' . $e->getMessage() );
-					}
-				}
+		if ( $amount > 0 ) {
```

Comment:
If `$amount === 0` (i.e. subscription with free trial), instead of doing nothing we must create a `SetupIntent`, at least to attach the `Payment Method` to the `Customer` object. That's why subscriptions with free trials weren't working for me, I realized it while I was cleaning up the branch.

Reply:
Hopefully, once $0 amounts are handled with a `SetupIntent`, then #698 and subscriber initiated payment method changes should _just work_  too (slightly [more details](https://docs.woocommerce.com/document/subscriptions/develop/payment-gateway-integration/#section-17)).

Reply:
/cc @luizreis as I see you're assigned to https://github.com/Automattic/woocommerce-payments/issues/698

Reply:
Handled in #832